### PR TITLE
boot_serial rework

### PR DESCRIPTION
**Pull request by mkiiskila**
Serial bootloader size was too big, so
 - got rid of snprintf() reference
 - got rid of cborattr
And on Mynewt side of size reduction:
 - tuned kernel/os to not bring memory pool code
 - replace minimal console with a more direct way of accessing uart

On some platforms the watchdog keeps running post-reset, and mynewt users have a gpio reporting when waiting for serial DFU. Therefore changed the code to pass platform specific routines to do reads and writes from uart. This allows Mynewt side to do it's watchdog tickling, gpio toggling without polling common code.

Unittests on mynewt side are still broken, but they've been so even before. So I'll create another PR to address those later.

Signed-Off-By: Marko Kiiskila <marko@apache.org>

### CODE REVIEW

**Comment by nvlsianpu** [comment]
I have tested this patch against zephyr and nrf52840_pca10056 target. I observed that sometimes (30%) communication had hanged during `image upgrade' command execution:
```
sudo ./mcumgr image upload -e ~/codedev/mcuboot/signed-smp_svr.bin  -c serial_0 
 13.21 KiB / 137.79 KiB [=====>-----------------------------------------------]   9.58% 1m51sError: Timeout reading from serial connection
```
After that It was impossible to renew the communication.

**Comment by mkiiskila** [comment]
Just tried the zephyr bootloader build on nrf52:
[marko@IsMyLaptop:~]$ newtmgr -c AL01QIP7 image upload ~/src2/incubator-mynewt-blinky/bin/targets/slinky_nrf52/app/apps/slinky/slinky.img
 55.45 KiB / 56.07 KiB [===============================================]  98.89%Done

This is what happened when I had forgotten a minicom open with that same tty:
[marko@IsMyLaptop:~/src2/incubator-mynewt-blinky]$ newtmgr -c AL01QIP7 image upload ~/src2/incubator-mynewt-blinky/bin/targets/slinky_nrf52/app/apps/slinky/slinky.img
 2.66 KiB / 56.07 KiB [==>-----------------------------------------]   4.74% 14sError: NMP timeout

**Comment by mkiiskila** [comment]
And here's the same mcumgr tool:

[marko@IsMyLaptop:~/foo/mynewt-mcumgr-cli/mcumgr]$ go run ./mcumgr.go -c AL01QIP7 image upload ~/src2/incubator-mynewt-blinky/bin/targets/slinky_nrf52/app/apps/slinky/slinky.img
 55.45 KiB / 56.07 KiB [===============================================]  98.89%Done

**Comment by mkiiskila** [comment]
That is interesting. I now built the bootloader for nrf52840, and it seems ok on my 10056. I did 10 uploads in a row, some using newtmgr and some with mucmgr. I did not go via application in the middle, however. Just stayed within the bootloader.

**Comment by nvlsianpu** [comment]
@mkiiskila ^^ I was using quite big image of the smp_svr example 141 KB.

**Comment by mkiiskila** [comment]
It appears to work for me even with bigger images. I tried 148k and 164k few times both. I assume this works ok when you run this from master?

**Comment by mkiiskila** [comment]
BTW, i'm not too savvy with zephyr builds. So I did this to build this:
osboxes@osboxes:~/zephyr/mcuboot$ git diff
diff --git a/boot/zephyr/prj.conf b/boot/zephyr/prj.conf
index 50c721f..eb0e152 100644
--- a/boot/zephyr/prj.conf
+++ b/boot/zephyr/prj.conf
@@ -1,6 +1,6 @@
-CONFIG_CONSOLE_HANDLER=y
-CONFIG_SYS_LOG=y
-CONFIG_DEBUG=y
+CONFIG_CONSOLE_HANDLER=n
+CONFIG_SYS_LOG=n
+CONFIG_DEBUG=n
 CONFIG_SYSTEM_CLOCK_DISABLE=y
 
 CONFIG_MAIN_STACK_SIZE=10240
@@ -17,9 +17,11 @@ CONFIG_MBEDTLS_CFG_FILE="mcuboot-mbedtls-cfg.h"
 
 CONFIG_FLASH=y
 CONFIG_MPU_ALLOW_FLASH_WRITE=y
-
+CONFIG_UART_CONSOLE=n
+CONFIG_BOOT_MAX_LINE_INPUT_LEN=512
+CONFIG_MCUBOOT_SERIAL=y
+CONFIG_TINYCBOR=y
 CONFIG_MULTITHREADING=n
-
 ### Zephyr keeps turning on major subsystems by default that we don't want.
 # CONFIG_BT is not set
 # CONFIG_I2C is not set

**Comment by nvlsianpu** [comment]
@mkiiskila 
>I assume this works ok when you run this from master?

Yes - I compare this against master - With master I was able to proceed like 10 times in row without issue.

**Comment by nvlsianpu** [comment]
@mkiiskila - Were you able to reproduce issue I had reported? I want to know whether have to reserve some additional time to investigate this by myslef?
cc @carlescufi - looks like I need to instigate the issue I signaled here https://github.com/runtimeco/mcuboot/pull/294#issuecomment-395373877

**Comment by carlescufi** [comment]
@nvlsianpu let's wait until @mkiiskila is able to reproduce first. Can you in the meantime post the details of your environment? Was this on Linux? On a Virtual Machine?

File: boot/mynewt/src/main.c [hunk 1]
```diff
@@ -23,28 +23,173 @@
...
 #include "bootutil/image.h"
 #include "bootutil/bootutil.h"
 #include "bootutil/bootutil_log.h"
 
-#define BOOT_AREA_DESC_MAX    (256)
-#define AREA_DESC_MAX         (BOOT_AREA_DESC_MAX)
+#if defined(MCUBOOT_SERIAL)
+#define BOOT_SERIAL_REPORT_DUR  \
+    (MYNEWT_VAL(OS_CPUTIME_FREQ) / MYNEWT_VAL(BOOT_SERIAL_REPORT_FREQ))
+#define BOOT_SERIAL_INPUT_MAX (512)
```
Comment (utzig) [inline boot/mynewt/src/main.c]: I believe this #define was added to boot_serial.c and can be removed from here.

**Comment by utzig** [comment]
I gave this a fair amount of testing today on the nrf52840_10056 with both Mynewt and Zephyr and had no issues at all. I think it's good to merge. @mkiiskila Could you squash the 2 review commits into the main commits, please!

**Comment by carlescufi** [comment]
This looks good to me, but I will wait for @nvlsianpu to give us feedback here on the issue he was seeing. Thanks @mkiiskila for this!

**Comment by carlescufi** [comment]
@mkiiskila I tried this again, and although I can now reproduce the timeout, this also happens on master. So we agree with @nvlsianpu that this PR can be merged.
@mkiiskila can you join the MCUboot slack channel? I'd like to ask you about how you are testing this.

**Pull request merged**